packages/router/src/route_definition/segment.rs

```diff
@@ -0,0 +1,484 @@
+    ///
+    /// # Panic
+    /// - If a _fixed_ route with the same `path` was already added, but only in debug builds.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use dioxus_router::prelude::*;
+    /// Segment::new().fixed("path", Route::new(()));
+    /// ```
+    pub fn fixed(mut self, path: &str, route: impl Into<Route>) -> Self {
+        if self.fixed.insert(path.to_string(), route.into()).is_some() {
+            error!(r#"two fixed routes with identical path: "{path}", later prevails"#);
+            #[cfg(debug_assertions)]
+            panic!(r#"two fixed routes with identical path: "{path}""#)
+        }
+
+        self
+    }
+
+    /// Add an _index_ route.
+    ///
+    /// The _index_ route acts like an `index.html` (with most web servers). It is active, if the
+    /// [`Segment`] is the first to not be specified by the path. Some examples:
+    /// - If the path is `/`, no segment is specified. The _index_ route of the root segment (the
+    ///   [`Segment`] passed to the [`Router`](crate::components::Router)) is active.
+    /// - If the path is `/test` or `/test/`, one segment is specified. If the root segment has an
+    ///   active route, and that route has a nested [`Segment`], that [`Segment`]s _index_  route is
+    ///   active.
+    ///
+    /// # Panic
+    /// - If an _index_ route was already set, but only in debug builds.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use dioxus_router::prelude::*;
+    /// Segment::new().index(());
+    /// ```
+    pub fn index(mut self, content: impl Into<RouteContent>) -> Self {
+        if !self.index.is_empty() {
+            error!("index route already set, later prevails");
+            #[cfg(debug_assertions)]
+            panic!("index route already set");
+        }
+
+        self.index = content.into();
+        self
+    }
+
+    /// Add a _matching_ parameter route.
+    ///
+    /// A _matching_ route allows the application to accept a dynamic value, that matches a provided
+    /// regular expression. A _matching_ parameter route is active if these conditions apply:
+    /// 1. The [`Segment`] is specified by the path,
+    /// 2. no _fixed_ route is active,
+    /// 3. no prior _matching_ route (in order of addition) is active,
+    /// 4. and the `regex` matches the segments value.
+    ///
+    /// The segments complete value will be provided as a parameter using the `key` specified in the
+    /// `route`.
+    ///
+    /// # URL decoding
+    /// - The segments value will be decoded when checking if the _matching_ route is active.
+    /// - The segments value will be decoded when providing it as a parameter.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use dioxus_router::prelude::*;
+    /// # use regex::Regex;
+    /// Segment::new().matching(Regex::new(".*").unwrap(), ParameterRoute::new("key", ()));
+    /// ```
+    pub fn matching(mut self, regex: Regex, route: impl Into<ParameterRoute>) -> Self {
```

Comment:
Oh, that's a very good idea. I'll try that.

Reply:
I just experimented with this approach a bit, and I'll definitely add a trait like this. However, I think it's preferable to keep the implementations of fixed routes and the parameter (or catch all) route separate.

The current ordering is:
1. fixed routes
2. matching routes
3. parameter (catch all)

With your suggestion, the parameter (catch all) route would be part of the matching routes, and might shadow other matching routes added later. Also, I'd like to keep fixed routes separate, because they are more specific than a pattern.

Reply:
I implemented your suggestion with the mentioned modification. Feel free to try it out. I'll add a documentation chapter a bit later.

I'm also thinking about adding an implementation for `String`, that supports basic wildcards or something similar. But I think I'll leave that for later PRs.

Reply:
I added an example to the traits documentation. I also added a short explanation to the matching route feature chapter.